Software/Program/Code/Core/Src/main.c

```diff
@@ -90,18 +147,121 @@ int main(void)
   /* Initialize all configured peripherals */
   MX_GPIO_Init();
   MX_TIM2_Init();
+  MX_ADC3_Init();
   /* USER CODE BEGIN 2 */
 
+   HAL_TIM_Base_Start_IT(&htim2);
+
   /* USER CODE END 2 */
 
   /* Infinite loop */
   /* USER CODE BEGIN WHILE */
-  while (1)
-  {
+   while (1)
+   {
+      static enum Direction_E Direction = DIRECTION_NOT_AVAILABLE;
+      static bool UpdateOutput = false;
+      static bool ADC_ConversionInProgress = false;
+      static uint16_t Previous_ADCValue = 0u;
+      static float Frequency = 0.0f;
+
+      /*** HANDLE FRQ_OUT GPIO ***/
+      
+
+      /*** HANDLE USER INPUT FOR HIGH-PULSE TIME ***/
+      // Check if button was pressed...
+      if ( ButtonPressed == true )
+      {
+         ButtonPressed = false;
+         HAL_Delay( DEBOUNCE_WAIT_TIME );  // TODO: Handle debounces without a delay... --> Suggestion: Use another timer and a flag in the ISR that prevents immediate setting of flag...
+
+         // Toggle high-pulse time amounts
+         Direction = (enum Direction_E) ( ( (int)Direction + 1 ) % ( (int)DIRECTION_ERROR + 1 ) );
+         UpdateOutput = true;
+      }
+
+      /*** HANDLE USER INPUT FOR FREQUENCY ***/
+      // Read ADC value and update timer auto-reload register (ARR) value to set new period
+      uint32_t adc_reading = 0;
+
+      // Start ADC conversion if one is not started already...
+      if ( ADC_ConversionInProgress == false )
+      {
+         (void)HAL_ADC_Start( &hadc3 );   // TODO: Use status return value
+         ADC_ConversionInProgress = true;
+         HAL_ADC_PollForConversion(&hadc3, 80);
+      //}
+      //else if ( ADC_ConversionComplete == true )
+      //{
+         ADC_ConversionComplete = false;
+         ADC_ConversionInProgress = false;
+         adc_reading = HAL_ADC_GetValue( &hadc3 );
+         if ( adc_reading != Previous_ADCValue )
+         {
+            UpdateOutput = true;
+            Previous_ADCValue = adc_reading;
+         }
+      }
+
+      /*** HANDLE FREQUENCY OUTPUT ***/
+      if ( UpdateOutput == true )
+      {
+         uint32_t high_pulse_time_arr_value;
+         float period_us;
+         uint32_t period_in_timer_counts;
+         
+         UpdateOutput = false;
+
+         // Convert this ADC reading into a frequency...
+         Frequency = Interpolate_ADCCount_To_Frq( adc_reading );
+         period_us = 1.0e9 / (float)Frequency;
+         period_in_timer_counts = (uint32_t) ( period_us * TIMER_COUNTS_PER_NANOSECOND );
```

Comment:
This should be with the ADC → `user_frequency_setpoint` section upstream, not here. This section can simply be:
```c
if ( true == UpdateOutput )
{
   // Update ARR registers
}
```